qemu/tests/vioinput_mice.py

```diff
@@ -0,0 +1,249 @@
+    :param wait_time: wait event received in listener event queue.
+    :param count: wheel event counts, default count=1.
+    """
+
+    scrolls = params.objects("scrolls")
+    exp_events = {'wheel-up': {'hScroll': 0, 'type': 'WHEELFORWARD'},
+                  'wheel-down': {'hScroll': 0, 'type': 'WHEELBACKWARD'}}
+    for scroll in scrolls:
+        error_context.context("Scroll mouse %s" % scroll, logging.info)
+        if "up" in scroll:
+            console.scroll_forward(count)
+        else:
+            console.scroll_backward(count)
+
+        events_queue = listener.events
+        time.sleep(wait_time)
+        error_context.context("Check correct scroll event is received",
+                              logging.info)
+        while not events_queue.empty():
+            for i in range(count):
+                event = events_queue.get()
+                # some windows os will return pointer move event first
+                # before return scroll event, so filter them here.
+                if event["type"] == "POINTERMOVE":
+                    continue
+                if event["type"] != exp_events[scroll]["type"]:
+                    test.fail("Received scroll events don't match expected events"
+                              "Received scroll events as: %s\n Expected events as: %s"
+                              % (event, scroll))
+
+
+@error_context.context_aware
+def mouse_move_test(test, params, console, listener,
+                    wait_time, end_pos, _abs):
+    """
+    Mouse move test.
+
+    :param test: kvm test object
+    :param params: Dictionary with the test parameters
+    :param console: graphical console.
+    :param listener: listening the mouse button event in guest.
+    :param wait_time: wait event received in listener event queue.
+    :param end_pos: a tuple of mouse destination position.
+    :param mtype: Mouse move type, abs or rel.
+    """
+
+    move_rate = int(params.get("move_rate", 80))
+    move_duration = int(params.get("move_duration", 1))
+    line = graphical_console.uniform_linear(move_duration, move_rate)
+    width, height = console.screen_size()
+    events_queue = listener.events
+    event_lst = []
+    start_pos = console.get_pointer_pos()
+    x0, y0 = start_pos
+    xn, yn = end_pos
+    # Compute a line y=kx+b through start_pos and end_pos.
+    if (xn - x0) != 0:
+        vertical = 0
+        k = (yn - y0) / (xn - x0)
+        b = yn - (k * xn)
+    else:
+        vertical = 1
+
+    error_context.context("Moving pointer from %s to %s" % (start_pos, end_pos))
+    console.pointer_move(end_pos, motion=line, _abs=_abs)
+    time.sleep(move_duration + wait_time)
+
+    error_context.context("Collecting all pointer move events from guest")
+    while not events_queue.empty():
+        event = events_queue.get()
+        xpos, ypos = event["xPos"], event["yPos"]
+        # Filter beyond screen size events.
+        # Due to os will ignores/corrects these events.
```

Comment:
Is there any official document pointed out this? if there is we can leave the link as a reference at here.

Reply:
hi ,xuhan, that's sorry, we have not the doc for that, I confirmed before with vioinput developer through mail.